### PR TITLE
fix: add support for mongodb driver v4

### PR DESCRIPTION
**Pull request by MatheusMeloAntiquera**
This new fix allow support to mongodb driver v4

Closes: #7907

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/entity-manager/MongoEntityManager.ts [hunk 1]
```diff
@@ -87,25 +100,51 @@ export class MongoEntityManager extends EntityManager {
...
     /**
-     * Finds entities that match given conditions.
+     * Finds entities that match given find options or conditions.
      */
-    async findBy<Entity>(
+    async find<Entity>(
         entityClassOrName: EntityTarget<Entity>,
-        where: any,
+        optionsOrConditions?: FindManyOptions<Entity> | Partial<Entity>,
     ): Promise<Entity[]> {
```
Comment (ezracelli) [inline src/entity-manager/MongoEntityManager.ts:109]: to match `mongodb`'s types, should the return type here be `Promise<WithId<Entity>[]>`? same goes for all the `MongoRepository.find*` and `MongoManager.find*` methods + `MongoRepository.createEntityCursor`

https://github.com/mongodb/node-mongodb-native/blob/d55022bfadfbddb16bf54bd9a801d901e0af99a0/src/collection.ts#L761

File: src/find-options/FindOneOptions.ts [hunk 2]
```diff
@@ -24,7 +18,7 @@ export interface FindOneOptions<Entity = any> {
     /**
      * Specifies what columns should be retrieved.
      */
-    select?: FindOptionsSelect<Entity> | FindOptionsSelectByString<Entity>
+    select?: FindOptionsSelect<Entity>
```
Comment (pleerock) [inline src/find-options/FindOneOptions.ts:21]: this is a breaking change. You should not remove this type signature, you'll break other people code (including people who use RDBMSes)

File: src/find-options/FindOneOptions.ts [hunk 3]
```diff
@@ -34,7 +28,7 @@ export interface FindOneOptions<Entity = any> {
     /**
      * Indicates what relations of entity should be loaded (simplified left join form).
      */
-    relations?: FindOptionsRelations<Entity> | FindOptionsRelationByString
+    relations?: FindOptionsRelations<Entity>
```
Comment (pleerock) [inline src/find-options/FindOneOptions.ts:31]: same here.

File: src/entity-manager/EntityManager.ts [hunk 4]
```diff
@@ -1291,7 +1291,7 @@ export class EntityManager {
     ): Repository<Entity> {
         // find already created repository instance and return it if found
         const repository = this.repositories.find(
-            (repository) => repository.target === target,
+            (repo) => repo.target === target,
```
Comment (pleerock) [inline src/entity-manager/EntityManager.ts:1294]: redundant change. please remove it.

**Comment by pleerock** [comment]
Overall it seems good, you just need to remove the changes I asked. Plus we need community (mongodb-users to add approvals to this PR).

File: package.json [hunk 5]
```diff
@@ -98,7 +98,8 @@
     "@types/js-yaml": "^4.0.0",
     "@types/mkdirp": "^1.0.1",
     "@types/mocha": "^8.2.1",
-    "@types/node": "^14.14.31",
+    "@types/mongodb": "^4.0.7",
```
Comment (Leichtwind) [inline package.json:101]: Doesn't mongodb@4 have built-in typings?

File: src/entity-manager/EntityManager.ts (hunk 4 above)
Comment (MatheusMeloAntiquera) [reply src/entity-manager/EntityManager.ts:1294]: There is a variable on 1293 line with the samename and and this was breaking the test

File: package.json (hunk 5 above)
Comment (MatheusMeloAntiquera) [reply package.json:101]: Maybe, I can try to remove the @types

**Comment by AtillaPehlivan** [comment]
Hi, i think `remove` and `removeOne` method should be changing with `delete` and `deleteOne` in the mongodb `FindOperatorsUnordered` interface becase of this https://jira.mongodb.org/browse/NODE-2978
i got some error today when i use initializeUnorderedBulkOp  using with `removeOne`

**Review by SouchetJulie — CHANGES_REQUESTED** [review]
~~You should update the l.254 of the README to reflect the correct version of mongodb ;)~~
Thank you :)

File: src/find-options/FindOneOptions.ts (hunk 2 above)
Comment (MatheusMeloAntiquera) [reply src/find-options/FindOneOptions.ts]: Done 👍

**Review by SouchetJulie — APPROVED** [review]
Seems good to me. Thanks for your work

**Comment by MatheusMeloAntiquera** [comment]
@SouchetJulie
I'm glad to help, thank you for your suggestion

**Comment by MatheusMeloAntiquera** [comment]
@AtillaPehlivan done 👍

**Comment by oktapodia** [comment]
@pleerock @imnotjames any news about merging this?

**Comment by pleerock** [comment]
Community said to merge it, we'll merge it 😆  Thanks for the contribution @MatheusMeloAntiquera

**Comment by pleerock** [comment]
I have to revert this PR because I found the issue. You are using imports `from "mongodb"`. We cannot use it because that way we'll have to force users to install mongodb typings (e.g. `npm i @types/mongodb`) when people don't need it (e.g. they use any other SQL database). Solution: copy/paste typings from @types/mongodb package into `src/driver/mongodb/typings.ts`. I know solution isn't ideal but we have to do it that way for now.

**Comment by pleerock** [comment]
Please create a new PR with the required changes. Thank you.

**Comment by MatheusMeloAntiquera** [comment]
> I have to revert this PR because I found the issue. You are using imports `from "mongodb"`. We cannot use it because that way we'll have to force users to install mongodb typings (e.g. `npm i @types/mongodb`) when people don't need it (e.g. they use any other SQL database). Solution: copy/paste typings from @types/mongodb package into `src/driver/mongodb/typings.ts`. I know solution isn't ideal but we have to do it that way for now.

Okay, I'll do it

**Comment by pleerock** [comment]
thanks... and btw I just noticed 5.0.0 version was released by mongodb, maybe directly add support for v5?

**Comment by MatheusMeloAntiquera** [comment]
> I have to revert this PR because I found the issue. You are using imports `from "mongodb"`. We cannot use it because that way we'll have to force users to install mongodb typings (e.g. `npm i @types/mongodb`) when people don't need it (e.g. they use any other SQL database). Solution: copy/paste typings from @types/mongodb package into `src/driver/mongodb/typings.ts`. I know solution isn't ideal but we have to do it that way for now.

Actually, mongodb package not need @types/mongodb anymore. Look: 
image
https://www.npmjs.com/package/@types/mongodb

So, Is it necessary to keep using `src/driver/mongodb/typings.ts`?

**Pull request merged**